presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -357,24 +357,26 @@ private TypeCompatibility typeCompatibilityForRow(RowType firstType, RowType sec
         }
 
         ImmutableList.Builder<RowType.Field> fields = ImmutableList.builder();
+        boolean firstCanCoerceToSecond = true;
         for (int i = 0; i < firstFields.size(); i++) {
-            Optional<Type> commonParameterType = getCommonSuperType(firstFields.get(i).getType(), secondFields.get(i).getType());
-            if (!commonParameterType.isPresent()) {
+            Type firstFieldType = firstFields.get(i).getType();
+            Type secondFieldType = secondFields.get(i).getType();
+            TypeCompatibility typeCompatibility = compatibility(firstFieldType, secondFieldType);
+            if (!typeCompatibility.isCompatible()) {
                 return TypeCompatibility.incompatible();
             }
+            Type commonParameterType = typeCompatibility.getCommonSuperType();
 
             Optional<String> firstParameterName = firstFields.get(i).getName();
             Optional<String> secondParameterName = secondFields.get(i).getName();
+            Optional<String> commonName = firstParameterName.equals(secondParameterName) ? firstParameterName : Optional.empty();
 
-            Optional<String> name = Optional.empty();
-            if (firstParameterName.equals(secondParameterName)) {
-                name = firstParameterName;
-            }
-
-            fields.add(new RowType.Field(name, commonParameterType.get()));
+            // ignore parameter name for firstCanCoerceToSecond
+            firstCanCoerceToSecond &= typeCompatibility.isFirstCanCoerceToSecond();
+            fields.add(new RowType.Field(commonName, commonParameterType));
         }
 
-        return TypeCompatibility.implicit(RowType.from(fields.build()), secondType);
+        return TypeCompatibility.explicit(RowType.from(fields.build()), firstCanCoerceToSecond);
```

Comment:
Not sure I understand what "explicit" means in this context.

Reply:
Maybe you can suggest a better name.

Here, `implicit` means `canCoerce` is set implicitly, based on whether `secondType` is equal to the `commonSuperType`. `explicit` means `canCoerce` is set explicitly.

Reply:
Oh, I see. So the "implicit" method just says leave it up to the TypeCompatibility object to determine, while "explicit" means the caller is making the choice. Yeah, those names are not ideal. Let me think about something better.